bin/propolis-server/src/lib/migrate/source.rs

```diff
@@ -83,16 +200,74 @@ struct SourceProtocol<T: AsyncRead + AsyncWrite + Unpin + Send> {
 
     /// Transport to the destination Instance.
     conn: WebSocketStream<T>,
+
+    /// Guest page table dirty bits to restore in the event of a migration
+    /// failure, so that a subsequent migration can attempt to offer only dirty
+    /// pages.
+    ///
+    /// Yes, this representation may seem somewhat weird. Bear with me: when we
+    /// call the `VM_NPT_OPERATION` ioctl, the pages whose dirty bits we would
+    /// like to set are represented by a bitmap. Therefore, we need to be able
+    /// to produce bitmaps for each chunk of guest physical address space for
+    /// which we've previously observed dirty pages. We could, of course, just
+    /// store a map of base guest addresses to 4096-byte arrays, keeping the
+    /// whole bitmap in memory until the migration finishes, but this seems kind
+    /// of sad, given that the bitmap may be fairly sparse.
+    ///
+    /// Instead, we store a map of base guest addresses to a list of indices
+    /// into the bitmap array and the byte that needs to go there. Then, if we
+    /// actually call `VM_NPT_OPERATION` to re-dirty those pages, we can go back
+    /// and bit-or the bytes at each index that we care about with the dirty
+    /// flags that we had previously cleared. This sparser representation should
+    /// save some memory versus storing a bunch of page-sized byte arrays.
+    ///
+    /// This is `Some` if we should attempt to re-set dirty bits on guest pages
+    /// in the event of a migration failure, and `None` if we cannot do so. It
+    /// will be `Some` if (and only if):
+    ///
+    /// - The current bhyve version supports the `VM_NPT_OPERATION` ioctl (i.e.
+    ///   it is at least bhyve v17 or later),
+    /// - A previous attempt to re-dirty pages has failed (viz.
+    ///   [`HAS_REDIRTYING_EVER_FAILED`]), in which case, we can no longer trust
+    ///   that all previously dirtied pages were re-dirtied correctly.
+    ///
+    /// Otherwise, we must fall back to always offering all pages in the initial
+    /// pre-pause RAM push phase.
+    dirt: Option<std::collections::HashMap<GuestAddr, Vec<(usize, u8)>>>,
 }
 
+const PAGE_BITMAP_SIZE: usize = 4096;
+
+/// Set if we were unable to re-set dirty bits on guest pages after a failed
+/// migration attempt. If this occurs, we can no longer offer only dirty pages
+/// in a subsequent migration attempt, as some pages which should be marked as
+/// dirty may not be.
+static HAS_REDIRTYING_EVER_FAILED: AtomicBool = AtomicBool::new(false);
```

Comment:
Why make this a static, and not, say, an attribute on the machine or something like that?

Reply:
My thinking was that, because it's state related to the migration source implementation in particular, it ought to be scoped to this module, rather than leaked out into the machine type. But, you're right that the static is a bit weird — even though a `propolis-server` process will only ever manage one VM, this module isn't really written with that assumption, and instead stores all the state on the machine type. I'm happy to move this if you prefer!

Reply:
Yeah, that makes sense. I'll add a struct in this module for migration-specific bits and add it to the machine state. Thanks!